### DISCUSSION TITLE
能否比较一下 Litepal 和 Room 两个 ORM 框架有什么不同的特点？

### DISCUSSION THREAD

**Discussion by amoureux-keeping** [Julius-lq/javaAnswer#3] [Q&A] [answered]
如题，能否比较一下 Litepal 和 Room 两个 ORM 框架有什么不同的特点？

**Comment by amoureux-keeping** [comment]
以下摘自郭神litapal仓库的issue：
“这两个库的差别点非常大，反而相似点很少，所以让我比较的话，我可能能比较出一大堆东西。

LitePal的API设计比较受ActiveRecord的影响，在数据库接口方面注重的是使用的简易性。
Room的API设计比较受Retrofit的影响，不提供便捷的数据库接口调用，更注重的是App架构的设计。

LitePal在尽可能地让开发者不写SQL，为此提供了一系列的便捷API。
Room需要开发者写很多SQL，但都以注解的形式封装到了DAO函数上面，这样当开发者调用函数时其实就是在执行SQL。

LitePal几乎不依赖于注解，少数注解也是用的运行时注解，功能比较简单。
Room高度依赖于注解，并且是通过编译时注解生成代码，性能更高。

LitePal提供了非常便利的表管理功能，升级表完全自动化，开发者只需要改一下版本号即可。
Room的表升级简直是噩梦，完全需要自己手写表升级语句，出错率极高。（因为第一代表是Room自动生成的，后续升级的话需要手写SQL语句，但是升级的SQL要和第一代表的结构完全一致，这块不用心写非常容易出错）

Room不提供异步操作接口，但是要求所有数据库操作在IO线程中进行。不过新版Room已支持协程，简化了异步的操作。
LitePal虽然提供了异步接口，但是在最新版本中已废弃，准备接下来参考Room，支持协程。

其他区别还有很多很多，我就不一一列举了。

总体上来说，Room的设计比较精简，非常清晰地告诉你我就三块东西，Database、Dao、Entity，这些功能就够你用的了。

而LitePal可能是因为历史原因吧，设计比较庞大，表管理系统就是一大块，还有增删改查，聚合函数等等功能，这些在Room看来都是不会提供的，给你一个@Query注解就都能解决了，SQL语句自己写去吧。

最后，LitePal我会一直维护下去，Room确实有很多设计得比较出色且值得借鉴的地方，我准备在接下来的几个版本中慢慢加进去。但是反过来，Room肯定是不会参考LitePal的，虽然我真的觉得Room的表升级方面得好好简化一下。”

**Comment by Julius-lq** [comment, marked as answer]
以下摘自郭神litapal仓库的issue：
“这两个库的差别点非常大，反而相似点很少，所以让我比较的话，我可能能比较出一大堆东西。

LitePal的API设计比较受ActiveRecord的影响，在数据库接口方面注重的是使用的简易性。
Room的API设计比较受Retrofit的影响，不提供便捷的数据库接口调用，更注重的是App架构的设计。

LitePal在尽可能地让开发者不写SQL，为此提供了一系列的便捷API。
Room需要开发者写很多SQL，但都以注解的形式封装到了DAO函数上面，这样当开发者调用函数时其实就是在执行SQL。

LitePal几乎不依赖于注解，少数注解也是用的运行时注解，功能比较简单。
Room高度依赖于注解，并且是通过编译时注解生成代码，性能更高。

LitePal提供了非常便利的表管理功能，升级表完全自动化，开发者只需要改一下版本号即可。
Room的表升级简直是噩梦，完全需要自己手写表升级语句，出错率极高。（因为第一代表是Room自动生成的，后续升级的话需要手写SQL语句，但是升级的SQL要和第一代表的结构完全一致，这块不用心写非常容易出错）

Room不提供异步操作接口，但是要求所有数据库操作在IO线程中进行。不过新版Room已支持协程，简化了异步的操作。
LitePal虽然提供了异步接口，但是在最新版本中已废弃，准备接下来参考Room，支持协程。

其他区别还有很多很多，我就不一一列举了。

总体上来说，Room的设计比较精简，非常清晰地告诉你我就三块东西，Database、Dao、Entity，这些功能就够你用的了。

而LitePal可能是因为历史原因吧，设计比较庞大，表管理系统就是一大块，还有增删改查，聚合函数等等功能，这些在Room看来都是不会提供的，给你一个@query注解就都能解决了，SQL语句自己写去吧。

最后，LitePal我会一直维护下去，Room确实有很多设计得比较出色且值得借鉴的地方，我准备在接下来的几个版本中慢慢加进去。但是反过来，Room肯定是不会参考LitePal的，虽然我真的觉得Room的表升级方面得好好简化一下。”
